include/cuco/bloom_filter.cuh

```diff
@@ -0,0 +1,393 @@
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   *
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt, class StencilIt, class Predicate>
+  __host__ void add_if(
+    InputIt first, InputIt last, StencilIt stencil, Predicate pred, cuda::stream_ref stream = {});
+
+  /**
+   * @brief Asynchronously adds keys in the range `[first, last)` if `pred` of the corresponding
+   * `stencil` returns `true`.
+   *
+   * @note The key `*(first + i)` is inserted if `pred( *(stencil + i) )` returns `true`.
+   *
+   * @tparam InputIt Device-accessible random access input iterator where
+   * <tt>std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * bloom_filter<K>::key_type></tt> is `true`
+   * @tparam StencilIt Device-accessible random-access iterator whose `value_type` is
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   *
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt, class StencilIt, class Predicate>
+  __host__ void add_if_async(InputIt first,
+                             InputIt last,
+                             StencilIt stencil,
+                             Predicate pred,
+                             cuda::stream_ref stream = {}) noexcept;
+
+  /**
+   * @brief Tests all keys in the range `[first, last)` if their fingerprints are present in the
+   * filter.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `test_async`.
+   *
+   * @tparam InputIt Device-accessible random access input iterator where
+   * <tt>std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * bloom_filter<K>::key_type></tt> is `true`
+   * @tparam OutputIt Device-accessible output iterator assignable from `bool`
+   *
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param output_begin Beginning of the sequence of booleans for the presence of each key
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt, class OutputIt>
+  __host__ void test(InputIt first,
```

Comment:
```suggestion
  __host__ void contains(InputIt first,
```
I prefer contains over test or find as it's the canonical term in C++ for a container method that returns a boolean query result. The fact that a bloom filter can generate false positives is common sense.

Reply:
I just did a quick survey how other libraries name their APIs.

- [Impala](https://github.com/apache/impala/blob/53a452af669b022fd2ab19eb58a5dc5ea29aed78/be/src/util/bloom-filter.h#L222): `Insert/Find`
- [Parquet](https://github.com/apache/arrow/blob/07420b0c56066326bd409e9537ee3d43ab6b1a51/cpp/src/parquet/bloom_filter.h#L46): `InsertHash/FindHash`
- [KVRocks](https://github.com/apache/kvrocks/blob/6e7476fb2cf9c20ab045ec54108c5877d0fca8d8/src/types/bloom_filter.h#L152): `InsertHash/FindHash`

Note that the last two expect the input to be a hash rather than a key. I wonder if this is something we want.

Reply:
Expecting (or at least supporting) a hash rather than a key has some nice characteristics in terms of re-using hash values that have already been pre-calculated. You can also prevent accidentally querying a filter created on hashes of (for instance) strings with a hash of a double (which leads to completely bogus results) by embedding the type you've hashed in a template parameter - a so called "phantom" parameter that is never used in the computations of the methods on the filter class itself, but which just prevents a specific type of footgun.

Reply:
> Expecting (or at least supporting) a hash rather than a key has some nice characteristics in terms of re-using hash values that have already been pre-calculated. 

Nice. I hadn’t considered it from that angle.

Reply:
Regarding the naming, `find` always returns an iterator in C++, so we should avoid using it here. I was torn between `insert` and `add` since `insert` is widely used in containers like `vector` and `unordered_map`, etc. However, a bloom filter doesn't "insert" data in the traditional sense; instead, it updates the filter to accommodate new data. I personally find `add` to be more appropriate, but `insert` would work as well.

Reply:
> re-using hash values that have already been pre-calculated

Yes, this could also be emulated by using `thrust::identity` for the hasher. 

> "phantom" parameter that is never used in the computation

I'm not sure I understand how this works in this case. If the `add/contains` functions only take a hash value, we cannot determine if the key that has been used to generate that hash value was equal to the phantom or not. Am I missing something here?

That said we could remove both the `Key` and `Hash` tparam from the class and expect raw hash values as function parameters and let the user deal with the proper generation of these hash values (we have an entire zoo of hash function ready to be used in cuco). It's probably one more line of code on the user side to generate the hash before passing it to the filter but that shouldn't be a dealbreaker.

Reply:
Quick survey: Any object on `add/contains` for the API names? If not, I would make the change then.

Reply:
`add`/`contains` +1

> That said we could remove both the Key and Hash tparam from the class

To clarify, I view `contains_hash` or `contains_hash_function` as an additional feature rather than a replacement for `contains`. Removing `Hash` and `Key` as template parameters seems more prone to errors.

Reply:
> The `add/contains` function themselves take an additional phantom parameter that can then be checked to verify the type. Is this correct?

Yep! I haven't tried this yet to see how ergonomic it is in different contexts, but that's the idea, anyway.

This is probably most useful when merging (unioning or intersecting) two filters, to avoid merging a filter created for doubles and one created for strings.